examples/fun-op-sem/listImp/listImpScript.sml

```diff
@@ -0,0 +1,239 @@
+
+open HolKernel Parse boolLib bossLib;
+open stringLib integerTheory;
+open listTheory;
+open arithmeticTheory;
+open impTheory;
+
+open optionTheory;
+
+val ect = BasicProvers.EVERY_CASE_TAC;
+
+val _ = new_theory "listImp";
+
+val _ = temp_type_abbrev("vname",``:string``);
+
+Datatype:
+  com = Assign vname imp$aexp
+    | If imp$bexp (com list) (com list)
+    | While imp$bexp (com list)
+End
+
+val com_size_def = fetch "-" "com_size_def";
+
+Definition pval_def:
+  (cval 0 _ _ = NONE) /\
+  (cval (t:num) (Assign v a) s = SOME ((v =+ aval a s) s)) /\
+  (cval t (If b cs1 cs2) s = pval t (if bval b s then cs1 else cs2) s) /\
+  (cval t (While b cs) s =
+    if bval b s then pval (t-1) (cs ++ [(While b cs)]) s
+    else SOME s) /\
+  (pval 0 _ _ = NONE) /\
+  (pval (t:num) [] s = SOME s) /\
+  (pval t (c::cs) s = case cval t c s of
+    | NONE => NONE
+    | SOME s' => pval t cs s')
+Termination
+  WF_REL_TAC `inv_image (measure I LEX measure I) (\r. case r of
+    | INR (t, cs, _) => (t, com1_size cs)
+    | INL (t, c, _) => (t, com_size c))` >>
+  rw[]
+End
+
+Definition imp_to_listImp_def:
+  (imp_to_listImp (SKIP:imp$com) = []) /\
+  (imp_to_listImp (Assign v a) = [((Assign v a):com)]) /\
+  (imp_to_listImp (Seq c1 c2) = (imp_to_listImp c1) ++ (imp_to_listImp c2)) /\
+  (imp_to_listImp (If b c1 c2) =
+    [(If b (imp_to_listImp c1) (imp_to_listImp c2))]) /\
+  (imp_to_listImp (While b c) = [(While b (imp_to_listImp c))])
+End
+
+Definition listImp_to_imp_def:
+  (prog_to_imp [] = SKIP) /\
+  (prog_to_imp (x::xs) = Seq (com_to_imp x) (prog_to_imp xs)) /\
+  (com_to_imp ((Assign v a):com) = ((Assign v a):imp$com)) /\
+  (com_to_imp (If b c1s c2s) = (If b (prog_to_imp c1s) (prog_to_imp c2s))) /\
+  (com_to_imp (While b cs) = (While b (prog_to_imp cs)))
+Termination
+  WF_REL_TAC `inv_image (measure I) (\r. case r of
+    | INR c => com_size c
+    | INL cs => com1_size cs)`
+End
+
+Theorem com_lt:
+  !c h t. MEM c t ==> com_size c < com1_size (h::t)
+Proof
+  rw[] >>
+  simp[com_size_def] >>
+  Induct_on `t` >>
+  rw[] >>
+  fs[com_size_def]
+QED
+
+Theorem com_leq:
```

Comment:
Proof should be `metis_tac[arithmeticTheory.LE_LT, com_lt]`

Reply:
This didn't end up working

```
metis: r[+0+7]+0+0+0+0+1+0+0+0!

Exception raised at folTools.FOL_FIND:
no solution found
Exception-
   HOL_ERR
     {message = "no solution found", origin_function = "FOL_FIND",
      origin_structure = "folTools"} raised
```